src/linear_algebra/general_linear_group.lean

```diff
@@ -0,0 +1,185 @@
+import linear_algebra.matrix.nonsingular_inverse
+import linear_algebra.special_linear_group
+import linear_algebra.determinant
+
+/-!
+# The General Linear group $GL(n, R)$
+
+This file defines the elements of the General Linear group `general_linear_group n R`,
+consisting of all invertible `n` by `n` `R`-matrices.
+
+
+## Main definitions
+
+ * `matrix.general_linear_group` is the type of matrices over R which are units in the matrix ring.
+ * `matrix.GL_plus.GL_pos` gives the subgroup of matrices with
+ positive determinant (over a linear ordered ring)
+
+## Implementation notes
+
+
+## References
+
+
+## Tags
+
+matrix group, group, matrix inverse
+-/
+
+namespace matrix
+universes u v
+open_locale matrix
+open linear_map
+
+section
+
+variables (n : Type u) [decidable_eq n] [fintype n] (R : Type v) [comm_ring R]
+
+/-- `GL n R` is the group of `n` by `n` `R`-matrices with unit determinant.
+Defined as a subtype of matrices-/
+abbreviation general_linear_group (R : Type*) [comm_ring R] : Type* := units (matrix n n R)
+
+notation `GL` := general_linear_group
+
+lemma GL.is_unit_det  (A : GL n R) : is_unit ((A : matrix n n R).det) :=
+begin
+  rw ←matrix.is_unit_iff_is_unit_det,
+  exact units.is_unit _
+end
+
+end
+namespace general_linear_group
+
+variables {n : Type u} [decidable_eq n] [fintype n] {R : Type v} [comm_ring R]
+
+/--The `GL n R` and `general_linear_group R n` groups are multiplicatively equivalent-/
+def to_lin : (GL n R) ≃* (linear_map.general_linear_group R (n → R)) :=
+units.map_equiv to_lin_alg_equiv'.to_mul_equiv
+
+/--Given a matrix with invertible determinant we get an element of `GL n R`-/
+def mk' (A : matrix n n R) (h : invertible (det A)) : GL n R :=
+unit_of_det_invertible A
+
+/--Given a matrix with unit determinant we get an element of `GL n R`-/
+noncomputable def mk'' (A : matrix n n R) (h : is_unit (det A)) : GL n R :=
+nonsing_inv_unit A h
+
+instance coe_fun : has_coe_to_fun (GL n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+lemma ext_iff (A B : GL n R) : A = B ↔ (∀ i j, (A : matrix n n R) i j = (B : matrix n n R) i j) :=
+units.ext_iff.trans matrix.ext_iff.symm
+
+/-- Not marked `@[ext]` as the `ext` tactic already solves this. -/
+lemma ext ⦃A B : GL n R⦄ (h : ∀ i j, (A : matrix n n R) i j = (B : matrix n n R) i j) :
+  A = B :=
+units.ext $ matrix.ext h
+
+/-- Applying `matrix.nonsing_inv` to the coercion is the same as coercing the group inverse. -/
+lemma nonsing_inv_coe_eq_coe_inv (A : GL n R) : (↑A : matrix n n R)⁻¹ = ↑(A⁻¹) :=
+begin
+  letI := A.invertible,
+  exact (inv_eq_nonsing_inv_of_invertible ↑A).symm,
+end
+
+section coe_lemmas
+
+variables (A B : GL n R)
+
+@[simp] lemma coe_fn_eq_coe : ⇑A = (↑A : matrix n n R) := rfl
+
+@[simp] lemma mul_val : ↑(A * B) = A ⬝ B := rfl
+
+@[simp] lemma one_val : ↑(1 : GL n R) = (1 : matrix n n R) := rfl
+
+@[simp] lemma inv_val : ↑(A⁻¹) = nonsing_inv A :=
+begin
+  have:=nonsing_inv_coe_eq_coe_inv A, unfold_coes at *, rw ← this, refl,
+end
```

Comment:
This is basically the same lemmas as `nonsing_inv_coe_eq_coe_inv`. Let's remove my lemma and prove this one directly:
```
@[simp] lemma coe_inv : ↑(A⁻¹) = (↑A : matrix n n R)⁻¹ :=
begin
  letI := A.invertible,
  exact inv_eq_nonsing_inv_of_invertible (↑A : matrix n n R),
end
```
Note we call this `coe_inv` not `inv_val`, because we're using `↑` not `.val`.

Reply:
Sure thing. I guess I should also remove `nonsing_inv_coe_eq_coe_inv` since it's the same lemma with the equality swapped.